salesforce/bulkwrite.go

```diff
@@ -0,0 +1,227 @@
+	location, err := url.JoinPath(baseURL, paths...)
+	if err != nil {
+		return "", errors.Join(err, common.ErrInvalidPathJoin)
+	}
+
+	return location, nil
+}
+
+func (c *Connector) createJob(ctx context.Context, config common.BulkWriteParams) (*ajson.Node, error) {
+	location, err := joinURLPath(c.BaseURL, "jobs/ingest")
+	if err != nil {
+		return nil, err
+	}
+
+	body := map[string]interface{}{
+		"object":              config.ObjectName,
+		"externalIdFieldName": config.ExternalId,
+		"contentType":         "CSV",
+		"operation":           "upsert",
+		"lineEnding":          "LF",
+	}
+
+	return c.post(ctx, location, body)
+}
+
+func (c *Connector) uploadCSV(ctx context.Context, jobId string, config common.BulkWriteParams) ([]byte, error) {
+	file, err := os.ReadFile(config.FilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload CSV for %s: %w", config.FilePath, errors.Join(err, common.ErrReadFile))
+	}
+
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/batches", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	return c.putCSV(ctx, location, file)
+}
+
+func (c *Connector) completeUpload(ctx context.Context, jobId string) (map[string]interface{}, error) {
+	updateLoadCompleteBody := map[string]interface{}{
+		"state": "UploadComplete",
+	}
+
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := c.patch(ctx, location, updateLoadCompleteBody)
+	if err != nil {
+		return nil, fmt.Errorf("patch failed: %w", errors.Join(err, common.ErrRequestFailed))
+	}
+
+	return ParseNodeToMap(res)
+}
+
+func (c *Connector) GetJobResult(ctx context.Context, jobId string) ([]byte, error) {
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/successfulResults", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	return c.getCSV(ctx, location)
+}
+
+func (c *Connector) GetUnprocessedResults(ctx context.Context, jobId string) ([]byte, error) {
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/unprocessedrecords", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	return c.getCSV(ctx, location)
+}
+
+func (c *Connector) GetAllJobs(ctx context.Context) (*ajson.Node, error) {
+	location, err := joinURLPath(c.BaseURL, "jobs/ingest")
+	if err != nil {
+		return nil, err
+	}
+
+	return c.get(ctx, location)
+}
+
+func (c *Connector) GetJobInfo(ctx context.Context, jobId string) (*ajson.Node, error) {
+	location, err := joinURLPath(c.BaseURL, "jobs/ingest", jobId)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.get(ctx, location)
+}
+
+func (c *Connector) FailedResults(ctx context.Context, jobId string) ([]byte, error) {
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/failedResults", jobId))
+	if err != nil {
+		return nil, err
+	}
+
+	return c.getCSV(ctx, location)
+}
+
+func ParseNodeToMap(node *ajson.Node) (map[string]interface{}, error) {
```

Comment:
ParseAjsonNodeToMap would be more clear (since this is an exported function, we should make its usage very clear)